### ISSUE TITLE
New rule: no-wildcard-export

### ISSUE THREAD

**Issue by gajus** [import-js/eslint-plugin-import#2493]
### ✅ Valid with `no-wildcard-export`

```ts
export { parseZodIssue } from './parseZodIssue';
```

### ❌ Invalid with `no-wildcard-export`

```ts
export * from './parseZodIssue';
```

## Why?

`export *` breaks static analysis of tools like [unimported](https://www.npmjs.com/package/unimported), effectively making it impossible to determine if there is unused code in the codebase.

**Comment by ljharb** [comment]
I'm not familiar with that tool, but I'm not sure why it would break static analysis - you'd just bounce over to the other file and see what _that_ exports.

**Comment by gajus** [comment]
> I'm not familiar with that tool, but I'm not sure why it would break static analysis - you'd just bounce over to the other file and see what that exports.

It is possible that it is a limitation of the tool. Nonetheless, it is caused by `export *`.

Don't want to muddy the waters, but I would also argue that `export *` hurts legibility of the codebase by obscuring the origin of exports. Whenever I clickthrough in VSCode to an `index.ts` file that only contains `export *` statements, it is near impossible to find where the declaration is coming from.

**Comment by ljharb** [comment]
I think if a compelling enough case can be presented, then yes - but i suspect such a case would also discourage any re-exporting, whether with `export *`, `export { x } from '…', or `import x, { y } from '…'; export default y; export { x }` etc.

**Comment by me4502** [comment]
IMO the main usecase for a rule like this over `no-restricted-syntax` is that it can have an auto-fixer when type information is available (replace the wildcard with all named exports), as well as build tooling performance. 

Especially in tests where modules are often isolated, knowing what exports are available from a file prevents having to parse that file when none of the exports are actually needed. This can substantially speed up build times on larger projects. 

While barrel files aren't best practice, they're still widely used across the industry and ecosystem. A lint rule like this would remove some of the pitfalls (duplicate exports, lack of clarity on where exports are coming from, etc), and improve build speeds potentially substantially.

**Comment by JounQin** [comment]
My question would be why `export *` is introduced into the ES spec if it should be considered as harmful, and I personally love this syntax because it's short and we don't need to manage exports a lot of places. The original issue should reported to `unimported` instead IMO.

**Comment by me4502** [comment]
I feel that's somewhat of a poor reason to not have a linting rule. Linting rules exist to match preferences of a codebase, and for codebases that want this it can have a major impact. 

While these seem to _break_ `unimported`, the fact that wildcard exports slow down _all_ build tools isn't something that build tools can avoid. They need to know if a file contains the export they're looking for, so they need to parse the file to determine that when a wildcard is used.

A rule like this that makes barrel files more usable and performant on larger codebases IMO makes a lot of sense. I don't believe this should be in the recommended set of linting rules, just one that people can enable themselves if they want.

**Comment by ljharb** [comment]
@JounQin quite a lot of things that are now considered harmful exist in the ES spec - existence in any spec is *never* an inherent endorsement for using anything. Explicit > implicit, and if you don't want to manage exports in a lot of places, *don't re-export them anywhere* - instead, deep import exactly what you need from the lone place it should be exported from.

@me4502 Barrel files are harmful - they cause tons of bundle/memory bloat that treeshaking can never fully account for, as well as conceptual and maintenance burden. They should not be made easier to use, they should be actively prevented.

**Comment by AntonNiklasson** [comment]
I'd like to enable a rule like this in our codebase. I'd like to enforce more explicit reexports in a few places, but we still find the barrel file pretty ergonomic. 

Are you open to contributions here? I could look into it, just want to make sure there's a green light.

**Comment by katryo** [comment]
I came to this thread from Google Search. I think `no-wildcard-export` rule makes sense for many developers.

Just fyi. As people suggested above, we can ban `export *` by `no-restricted-syntax` in this way.

```
"no-restricted-syntax": [
      "error",
      {
        selector: ":matches(ExportAllDeclaration)",
        message: "Export only modules you need.",
      },
    ],
```
